add "when to write rules" advice (#285)
JuliaDiff/ChainRulesCore.jl: docs/src/writing_good_rules.md

See docstrings for the comprehensive usage instructions.
## Write tests

[ChainRulesTestUtils.jl](https://github.com/JuliaDiff/ChainRulesTestUtils.jl)
provides tools for writing tests based on [FiniteDifferences.jl](https://github.com/JuliaDiff/FiniteDifferences.jl).
Take a look at the documentation or the existing [ChainRules.jl](https://github.com/JuliaDiff/ChainRules.jl) tests to see how to write the tests.

> It's easy!

Reviewer: ```suggestion
```

!!! warning
    Don't use analytical derivations for derivatives in the tests.
    Those are what you use to define the rules, and so can not be confidently used in the test.
    If you misread/misunderstood them, then your tests/implementation will have the same mistake.
    Use finite differencing methods instead, as they are based on the primal computation.

## CAS systems are your friends.

It is very easy to check gradients or derivatives with a computer algebra system (CAS) like [WolframAlpha](https://www.wolframalpha.com/input/?i=gradient+atan2%28x%2Cy%29).

## Which functions need rules?

In principle, a perfect AD system only needs rules for basic operations and can infer the

> rules for more complicated functions automatically.

Reviewer: minor style comment: in markdown docs, writing one sentence per line (rather than having a character limit for lines) makes it easier to both make suggestions and to review diffs in future PRs

In practice, performance needs to be considered as well.

Some functions use `ccall` internally for performance reasons, for example [`^`]

> (https://github.com/JuliaLang/julia/blob/380abc808ad8fbd72a7a57f6ac0f811b7903046d/base/math.jl#L922).

Reviewer: not sure it matters, but maybe pointing to a release version will seem less arbitrary and be just as future-proof
```suggestion
(https://github.com/JuliaLang/julia/blob/v1.5.3/base/math.jl#L886).
```

> These functions can not be differentiated through by AD systems, and need custom rules.

Author: Is this actually true? IRC Enzyme works at llvm level so might be able to do these?

Author: @oxinabox your wisdom is required

Reviewer 2: 2 points:
1.) It isn't normally for performance reasons. For a lot of mathematical functions its just that it hasn't been ported to julia yet (e.g. `FastMath`), because porting it is a lot of work for litttle gain. For a lot of other things, like allocating and resizing arrays, it is expressing a fundermental operation that doesn't exist in julia. Also all built-ins

2.) Enzyme *can* do these things, but it is a special case. We are talking at a general enough level. In general AD systems can not cross the language binaries, nor AD things that are too primative (fundermentally one at very least has to hit things that has assembly instructions, since there is nothing to decompose). We should not worry about being so strictly true that this becomes hard to read

Other functions can in principle be differentiated through by an AD system, but there is a
mathematical insight that can dramatically improve the computation of the derivative.
An example is numerical integration, where writing a rule removes the need to perform
AD through numerical integration.

Furthermore, AD systems make different trade-offs in performance due to their design.
This means that a certain rule will help one AD system, but not improve (and also not harm)
another.

> Below, we list some patterns relevant for the [`Zygote`](https://github.com/FluxML/Zygote.jl)

Reviewer: our style in docs seems to be to write e.g. "ChainRules.jl" rather than "`ChainRules`"

```suggestion
Below, we list some patterns relevant for the [Zygote.jl](https://github.com/FluxML/Zygote.jl)
```

AD system.

> ### Patterns that need rules in [`Zygote`](https://github.com/FluxML/Zygote.jl)

Reviewer: ```suggestion
### Patterns that need rules in [Zygote.jl](https://github.com/FluxML/Zygote.jl)
```

There are a few classes of functions that Zygote can not differentiate through.
Custom rules will need to be written for these to make AD work.

> Other patterns can be AD'ed through, but they are slow.

Author: ```suggestion
Other patterns can be AD'ed through, but the backward pass performance can be greatly improved by writing a rule.
```

#### Functions which mutate arrays
For example,
```julia
function addone!(array)
    array .+= 1
    return sum(array)
end

> addone!_cr(array) = addone!(array)

Reviewer: if we're keeping the `*_cr` functions, i think i'd move it down to where it's used, and add a comment (in addition to the text box below)

```suggestion
```

```
complains that
```julia
julia> using Zygote
julia> gradient(addone!, a)
ERROR: Mutating arrays is not supported
```

> However, upon adding the `rrule` (restart the REPL first)

Reviewer: does restarting the repl remove the need for the extra `addone!_cr` function? that'd simplify the example a lot...

Reviewer: alternatively, can it be deleted
```suggestion
However, upon adding the `rrule`
```

Author: Yes sorry forgot to delete this. I felt it's easier to follow with REPL if you have two functions rather than restarting REPL. But reading it might be more difficult.

Perhaps the best way is to have it more readable (with a comment to restart REPL, and a single function). I can change the box to reflect that.

Reviewer: up to you - i think either is fine (but both was confusing)

```julia

> function ChainRules.rrule(::typeof(addone!_cr), a)

Reviewer: ```suggestion
# Define a new function so it's a different `gradient` call and the `rrule` will be found.
addone!_cr(array) = addone!(array)

function ChainRules.rrule(::typeof(addone!_cr), a)
```

#### Exception handling

Zygote does not support differentiating through `try`/`catch` statements.
For example, differentiating through
```julia
function exception(x)
    try
        return x^2
    catch e
        println("could not square input")
        throw(e)
    end
end
exception_cr(x) = exception(x)
```
does not work
```julia
julia> gradient(exception, 3.0)
ERROR: Compiling Tuple{typeof(exception),Int64}: try/catch is not supported.
```
without an `rrule` defined
```julia
function ChainRulesCore.rrule(::typeof(exception_cr), x)
    y = exception_cr(x)
    function exception_cr_pullback(ȳ)
        return (NO_FIELDS, 2*x)
    end
    return y, exception_cr_pullback
end
```

```julia
julia> gradient(exception_cr, 3.0)
(6.0,)
```

#### Loops

Julia runs loops fast.
Unfortunately Zygote differentiates through loops slowly.
So, for example, computing the mean squared error by using a loop
```julia
function mse(y, ŷ)
    N = length(y)
    s = 0.0
    for i in 1:N
        s +=  (y[i] - ŷ[i])^2.0
    end
    return s/N
end
mse_cr(y, ŷ) = mse(y, ŷ)
```
takes a lot longer to AD through
```julia
julia> y = rand(30)
julia> ŷ = rand(30)

> julia> @btime gradient(mse, y, ŷ)

Reviewer: do we need to interpolate in these values for an accurate benchmark?

Author: yes indeed nice catch. Zygote really is bad with loops, I didn't realise just how bad it is

  39.561 μs (1007 allocations: 65.33 KiB)
```
than if we supply an `rrule`,
```julia
function ChainRules.rrule(::typeof(mse_cr), x, x̂)
    output = mse_cr(x, x̂)
    function mse_cr_pullback(ȳ)
        N = length(x)
        g = (2 ./ N) .* (x .- x̂) .* ȳ
        return (NO_FIELDS, g, -g)
    end
    return output, mse_cr_pullback
end
```
which is much faster
```julia
julia> @btime gradient(mse_cr, y, ŷ)
  1.293 μs (18 allocations: 1.09 KiB)
```

#### Inplace accumulation

Inplace accumulation of gradients is slow in `Zygote`. The issue, demonstrated in the
folowing example, is that the gradient of `getindex` allocates an array of zeros with
a single non-zero element. 
```julia

> function inplace(array)

Reviewer: i find it very confusing that this is named `inplace` when it doesn't do anything "inplace"

i think it's to match the heading of this section, which is about "Inplace accumulation", but i think it's more confusing that giving it a name like `sum3` or something that'd you'd actually name a function with this behaviour

    x = array[1]
    y = array[2]
    z = array[3]
    return x+y+z
end
inplace_cr(a) = inplace(a)
```
Computing the gradient with only a single array allocation using an `rrule`
```julia
function ChainRulesCore.rrule(::typeof(inplace_cr), a)
    y = inplace_cr(a)
    function inplace_pullback(ȳ)
        grad = zeros(length(a))
        grad[1:3] .+= 1.0
        return NO_FIELDS, grad
    end
    return y, inplace_pullback
end
```
turns out to be significantly faster

> ```

Reviewer: ```suggestion
```julia
```
